src/sync/mpsc/mod.rs

```diff
@@ -0,0 +1,736 @@
+                PopResult::Inconsistent => {
+                    // Same as above
+                    thread::yield_now();
+                }
+            }
+        }
+    }
+
+    // Try to park the receiver task
+    fn try_park(&self) -> TryPark {
+        let curr = self.inner.state.load(Ordering::SeqCst);
+        let state = decode_state(curr);
+
+        // If there are pending messages, then there is no need to park.
+        if state.num_messages > 0 {
+            return TryPark::NotEmpty;
+        }
+
+        // If the channel is closed, then there is no need to park.
+        if !state.is_open {
+            return TryPark::Closed;
+        }
+
+        // First, track the task in the `recv_task` slot
+        let mut recv_task = self.inner.recv_task.lock().unwrap();
+
+        if recv_task.unparked {
+            // Consume the `unpark` signal without actually parking
+            recv_task.unparked = false;
+            return TryPark::NotEmpty;
+        }
+
+        recv_task.task = Some(task::park());
+        TryPark::Parked
+    }
+
+    fn dec_num_messages(&self) {
+        let mut curr = self.inner.state.load(Ordering::SeqCst);
+
+        loop {
+            let mut state = decode_state(curr);
+
+            state.num_messages -= 1;
+
+            let next = encode_state(&state);
+            let actual = self.inner.state.compare_and_swap(curr, next, Ordering::SeqCst);
+
+            if actual == curr {
+                return;
+            }
+
+            curr = actual;
+        }
+    }
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<T>, ()> {
+        loop {
+            // Try to read a message off of the message queue.
+            let msg = match self.next_message() {
+                Async::Ready(msg) => msg,
+                Async::NotReady => {
+                    // There are no messages to read, in this case, attempt to
+                    // park. The act of parking will verify that the channel is
+                    // still empty after the park operation has completed.
+                    match self.try_park() {
+                        TryPark::Parked => {
+                            // The task was parked, and the channel is still
+                            // empty, return NotReady.
+                            return Ok(Async::NotReady);
+                        }
+                        TryPark::Closed => {
+                            // The channel is closed, there will be no further
+                            // messages.
+                            return Ok(Async::Ready(None));
+                        }
+                        TryPark::NotEmpty => {
+                            // A message has been sent while attempting to
+                            // park. Loop again, the next iteration is
+                            // guaranteed to get the message.
+                            continue;
```

Comment:
Yes, it can only loop a single time. It can only reach this point if it previously didn't pop a message, tried to park, and during the park a sender completed the send process (see the unparked flag in `try_park`.